sgkit/io/vcf/vcf_writer.py

```diff
@@ -0,0 +1,291 @@
+import re
+from typing import MutableMapping, Union
+
+import numpy as np
+
+from sgkit import load_dataset
+from sgkit.io.vcf.vcf_writer_utils import (
+    byte_buf_to_str,
+    create_mask,
+    interleave,
+    vcf_fixed_to_byte_buf,
+    vcf_fixed_to_byte_buf_size,
+    vcf_format_missing_to_byte_buf,
+    vcf_format_names_to_byte_buf,
+    vcf_format_names_to_byte_buf_size,
+    vcf_genotypes_to_byte_buf,
+    vcf_genotypes_to_byte_buf_size,
+    vcf_info_to_byte_buf,
+    vcf_info_to_byte_buf_size,
+    vcf_values_to_byte_buf,
+    vcf_values_to_byte_buf_size,
+)
+from sgkit.typing import PathType
+
+
+def zarr_to_vcf(
+    input: Union[PathType, MutableMapping[str, bytes]],
+    output: PathType,
+) -> None:
+    """Convert a Zarr file to VCF file.
+
+    Warnings
+    --------
+    This function requires the input Zarr file to have a ``vcf_header`` attribute
+    containing the VCF header. VCF files converted to Zarr using :func:`vcf_to_zarr`
+    will contain this attribute, but datasets loaded from other sources will not.
+
+    The current implementation loads the entire dataset into memory, so it should
+    only be used on suitably-sized datasets.
+
+    Parameters
+    ----------
+    input
+        Zarr store or path to directory in file system.
+    output
+        A path that the output VCF should be written to.
+    """
+
+    ds = load_dataset(input)
+    ds = ds.load()
+
+    header_str = ds.attrs["vcf_header"]
+
+    n_variants = ds.dims["variants"]
+    n_samples = ds.dims["samples"]
+
+    with open(output, mode="w") as out:
+        print(header_str, end="", file=out)
+
+        header_info_fields = _info_fields(header_str)
+        header_format_fields = _format_fields(header_str)
+
+        # fixed fields
+
+        contigs = np.array(ds.attrs["contigs"], dtype="S")
+        filters = np.array(ds.attrs["filters"], dtype="S")
+
+        chrom = ds.variant_contig.values
+        pos = ds.variant_position.values
+        id = ds.variant_id.values.astype("S")
+        alleles = ds.variant_allele.values.astype("S")
+        qual = ds.variant_quality.values
+        filter_ = ds.variant_filter.values
+
+        # info fields
+
+        # preconvert all info fields to byte representations
+        info_bufs = []
+        info_mask = np.full((len(header_info_fields), n_variants), False, dtype=bool)
+        info_indexes = np.zeros(
+            (len(header_info_fields), n_variants + 1), dtype=np.int32
+        )
+
+        k = 0
+        info_prefixes = []  # field names followed by '=' (except for flag/bool types)
+        for key in header_info_fields:
```

Comment:
Probably worth making ``info_field = f"variant_key"`` here, it's used 3 or 4 times